extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java

```diff
@@ -209,9 +218,21 @@ public boolean isValid(Connection connection) {
             }
         }
 
+        // Identify AgroalDataSourceListener beans and filter appropriately to this dataSourceName.
+        List<PrioritizedDataSourceListenerWrapper> listeners = beanManager.getBeans(AgroalDataSourceListener.class).stream()
```

Comment:
The problem with this approach is that beans implementing `AgroalDataSourceListener` would be normally removed since they're considered "unused" (see https://quarkus.io/guides/cdi-reference#remove_unused_beans). When using `javax.enterprise.inject.Instance`, on the other hand, the bean class is not available (DataSource name is - see my comment bellow).

Maybe we should introduce a quarkus-specific interface that would extend `AgroalDataSourceListener` and `javax.enterprise.inject.spi.Prioritized`. That would make it possible to simply `@Inject Instance<PrioritizedDataSourceListener> instance` then `select(DataSourceLiteral.of("dataSourceName"))` (DataSourceLiteral would be something similar to `javax.enterprise.inject.literal.NamedLiteral`) and then simply `instance.stream().sorted()...`.

Reply:
In my project using this, I have an intermediate library that has the listener, so I guess that's why it's not getting removed. Good to know.

I like the suggestion of the custom interface. I was trying to make this as small a change as possible, but I'd rather make it as correct a change as possible. :-)

I think the combined quarkus-specific thing is a good idea - it also provides a place to define some priority constants specific to this use.

Regarding testing, I didn't see any tests in the quarkus-agroal projects to base testing off of.
If you have a good crash-course on how to test quarkus stuff -- I'd love to see it. This has been nagging at me.

Reply:
So we should have a test in the deployment module: https://github.com/quarkusio/quarkus/tree/master/extensions/agroal/deployment/src/test and ideally also a test for native image, somewhere here: https://github.com/quarkusio/quarkus/tree/master/integration-tests.

Reply:
@mkouba - To implement the existing AgroalEventLoggingListener there'll need to be a build-time injection point transformer, correct?

Reply:
Also, will ARC automagically sort based on priority if things extend Prioritized?

Reply:
@mkouba could you answer the question above ^.

Reply:
Sorry, I lost the context and I will probably need a simplified summary...

In any case, ArC does not sort the beans. However, it's possible to inject `javax.enterprise.inject.Instance` and then use `instance.stream().sorted()`.